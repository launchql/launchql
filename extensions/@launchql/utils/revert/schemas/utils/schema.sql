-- Revert schemas/utils/schema from pg

BEGIN;

DROP SCHEMA utils;

COMMIT;
