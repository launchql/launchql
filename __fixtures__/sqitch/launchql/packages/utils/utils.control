# utils extension
comment = 'skitch project'
default_version = '0.0.1'
module_pathname = '$libdir/utils'
requires = 'plpgsql,uuid-ossp,totp,pg-verify'
relocatable = false
superuser = false
