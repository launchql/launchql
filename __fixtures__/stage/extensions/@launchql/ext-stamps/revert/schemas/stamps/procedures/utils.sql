-- Revert schemas/stamps/procedures/utils from pg

BEGIN;

DROP FUNCTION stamps.utils;

COMMIT;
