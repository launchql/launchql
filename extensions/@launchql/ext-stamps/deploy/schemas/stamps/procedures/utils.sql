-- Deploy schemas/stamps/procedures/utils to pg

-- requires: schemas/stamps/schema

BEGIN;

CREATE FUNCTION stamps.peoplestamps()
RETURNS TRIGGER AS $$
BEGIN
    IF TG_OP = 'INSERT' THEN
      NEW.created_by = jwt_public.current_user_id();
      NEW.updated_by = jwt_public.current_user_id();
    ELSIF TG_OP = 'UPDATE' THEN
      NEW.created_by = OLD.created_by;
      NEW.updated_by = jwt_public.current_user_id();
    END IF;
    RETURN NEW;
END;
$$ LANGUAGE 'plpgsql';

CREATE FUNCTION stamps.timestamps()
RETURNS TRIGGER AS $$
BEGIN
    IF TG_OP = 'INSERT' THEN
      NEW.created_at = NOW();
      NEW.updated_at = NOW();
    ELSIF TG_OP = 'UPDATE' THEN
      NEW.created_at = OLD.created_at;
      NEW.updated_at = NOW();
    END IF;
    RETURN NEW;
END;
$$ LANGUAGE 'plpgsql';

COMMIT;
