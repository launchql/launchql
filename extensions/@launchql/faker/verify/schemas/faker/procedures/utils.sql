-- Verify schemas/faker/procedures/utils  on pg

BEGIN;

SELECT verify_function ('faker.utils');

ROLLBACK;
