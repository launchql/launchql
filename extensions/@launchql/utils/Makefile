EXTENSION = launchql-utils
DATA = sql/launchql-utils--0.4.6.sql

PG_CONFIG = pg_config
PGXS := $(shell $(PG_CONFIG) --pgxs)
include $(PGXS)
  