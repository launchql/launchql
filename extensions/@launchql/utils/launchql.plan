%syntax-version=1.0.0
%project=launchql-utils
%uri=launchql-utils
  
schemas/utils/schema 2017-08-11T08:11:51Z skitch <skitch@5b0c196eeb62> # add schemas/utils/schema
schemas/utils/procedures/mask_pad [schemas/utils/schema] 2017-08-11T08:11:51Z skitch <skitch@5b0c196eeb62> # add schemas/utils/procedures/mask_pad
schemas/utils/procedures/throw [schemas/utils/schema] 2017-08-11T08:11:51Z skitch <skitch@5b0c196eeb62> # add schemas/utils/procedures/throw
schemas/utils/procedures/ensure_singleton [schemas/utils/schema] 2021-04-25T11:50:15Z Dan Lynch <dlynch@Dans-MBP-3> # add schemas/utils/procedures/ensure_singleton
