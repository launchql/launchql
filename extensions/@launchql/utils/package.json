{
  "_from": "@launchql/utils@^0.4.6",
  "_id": "@launchql/utils@0.4.6",
  "_inBundle": false,
  "_integrity": "sha512-1ZB6FYCWUCr0casWo7m8Xr74IGpwVSlaTXGnVhslXoLAaU/nwIr39khzRRkHKIrjJNG4VoWfMTymZLfDAQDj1Q==",
  "_location": "/@launchql/utils",
  "_phantomChildren": {},
  "_requested": {
    "type": "range",
    "registry": true,
    "raw": "@launchql/utils@^0.4.6",
    "name": "@launchql/utils",
    "escapedName": "@launchql%2futils",
    "scope": "@launchql",
    "rawSpec": "^0.4.6",
    "saveSpec": null,
    "fetchSpec": "^0.4.6"
  },
  "_requiredBy": [
    "/@pyramation/launchql-meta-schema"
  ],
  "_resolved": "https://registry.npmjs.org/@launchql/utils/-/utils-0.4.6.tgz",
  "_shasum": "4d30448b4bc2efa4891add0b9d393c66ad181333",
  "_spec": "@launchql/utils@^0.4.6",
  "_where": "/private/var/folders/kn/177lldss4n9fwn0gh5q5wfn00000gn/T/ny4sag54vjfki5oa3m947/extensions/node_modules/@pyramation/launchql-meta-schema",
  "author": {
    "name": "Dan Lynch",
    "email": "pyramation@gmail.com"
  },
  "bundleDependencies": false,
  "deprecated": false,
  "description": "utils",
  "devDependencies": {
    "@babel/cli": "7.10.4",
    "@babel/core": "7.10.4",
    "@babel/plugin-proposal-class-properties": "7.10.4",
    "@babel/plugin-proposal-export-default-from": "7.10.4",
    "@babel/plugin-proposal-object-rest-spread": "7.10.4",
    "@babel/plugin-transform-runtime": "7.10.4",
    "@babel/preset-env": "7.10.4",
    "@babel/runtime": "^7.4.2",
    "@launchql/cli": "1.3.5",
    "@launchql/db-testing": "latest",
    "babel-eslint": "10.1.0",
    "babel-jest": "26.1.0",
    "eslint": "^7.3.1",
    "eslint-config-prettier": "^6.10.0",
    "eslint-plugin-prettier": "^3.1.2",
    "jest": "26.1.0",
    "prettier": "2.0.5",
    "regenerator-runtime": "^0.13.2"
  },
  "gitHead": "7e686ca8bfee944da876b6cfd2c32e71da244537",
  "name": "@launchql/utils",
  "publishConfig": {
    "access": "public"
  },
  "scripts": {
    "prepare": "lql package",
    "test": "FAST_TEST=1 launchql-templatedb && jest",
    "test:watch": "FAST_TEST=1 jest --watch"
  },
  "version": "0.4.6"
}
