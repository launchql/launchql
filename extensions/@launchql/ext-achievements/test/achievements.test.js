import { getConnections } from './utils';
import { snap } from './utils/snaps';
let db, teardown;
const objs = {
  tables: {}
};
const user_id = 'b9d22af1-62c7-43a5-b8c4-50630bbd4962';
const status = {};

const repeat = (str, n) => Array.from({ length: n }).fill([str]).flat();

beforeAll(async () => {
  ({ db, teardown } = await getConnections());
  // await db.begin();
  // await db.savepoint();

  status.public = db.helper('status_public');
  status.private = db.helper('status_private');
});

afterAll(async () => {
  try {
    // try catch here allows us to see the sql parsing issues!
    // await db.rollback();
    // await db.commit();
    await teardown();
  } catch (e) {
    // noop
  }
});

afterEach(async () => {
  await db.afterEach();
});

const levels = ['newbie', 'advanced'];

const newbie = [
  ['upload_profile_image'],
  ['complete_action', 5],
  ['accept_cookies'],
  ['accept_privacy'],
  ['agree_to_terms']
];
const advanced = [
  ['invite_users', 15],
  ['complete_action', 15]
];

beforeEach(async () => {
  await db.beforeEach();
  //
  for (const name of levels) {
    await status.public.insertOne('levels', {
      name
    });
  }
  //
  for (const [name, required_count = 1] of newbie) {
    await status.public.insertOne('level_requirements', {
      name,
      level: 'newbie',
      required_count
    });
  }
  for (const [name, required_count = 1] of advanced) {
    await status.public.insertOne('level_requirements', {
      name,
      level: 'advanced',
      required_count
    });
  }
});

it('newbie', async () => {
  //

  const steps = [
    'agree_to_terms',
    'accept_cookies',
    'accept_privacy',
    ...repeat('complete_action', 3)
  ];
  for (const name of steps) {
    await status.public.insertOne('user_steps', {
      user_id,
      name
    });
  }

  const advancedRequirements = await status.public.callAny('steps_required', {
    level: 'advanced',
    user_id
  });
  snap({ advancedRequirements });
  const newbieRequirements = await status.public.callAny('steps_required', {
    level: 'newbie',
    user_id
  });
  snap({ newbieRequirements });

  const [userAchievedNewbie] = await status.public.callAny('user_achieved', {
    level: 'newbie',
    user_id
  });
  snap({ newbie: userAchievedNewbie });
  const [userAchievedAdvanced] = await status.public.callAny('user_achieved', {
    level: 'advanced',
    user_id
  });
  snap({ advanced: userAchievedAdvanced });
});

it('advanced', async () => {
  //

  const steps = [
    'agree_to_terms',
    'accept_cookies',
    'accept_privacy',
    'upload_profile_image',
    ...repeat('invite_users', 3),
    ...repeat('complete_action', 21)
  ];
  for (const name of steps) {
    await status.public.insertOne('user_steps', {
      user_id,
      name
    });
  }

  const advancedRequirements = await status.public.callAny('steps_required', {
    level: 'advanced',
    user_id
  });
  snap({ advancedRequirements });
  const newbieRequirements = await status.public.callAny('steps_required', {
    level: 'newbie',
    user_id
  });
  snap({ newbieRequirements });

  const [userAchievedNewbie] = await status.public.callAny('user_achieved', {
    level: 'newbie',
    user_id
  });
  snap({ newbie: userAchievedNewbie });
  const [userAchievedAdvanced] = await status.public.callAny('user_achieved', {
    level: 'advanced',
    user_id
  });
  snap({ advanced: userAchievedAdvanced });
});

it('advanced part II', async () => {
  //

  const advanced = [['apply_for_verifier'], ['create_action', 2]];

  for (const [name, required_count = 1] of advanced) {
    await status.public.insertOne('level_requirements', {
      name,
      level: 'advanced',
      required_count
    });
  }

  const steps = [
    'agree_to_terms',
    'accept_cookies',
    'accept_privacy',
    'upload_profile_image',
    ...repeat('invite_users', 3),
    ...repeat('complete_action', 10)
  ];
  for (const name of steps) {
    await status.public.insertOne('user_steps', {
      user_id,
      name
    });
  }

  const advancedRequirements = await status.public.callAny('steps_required', {
    level: 'advanced',
    user_id
  });
  snap({ advancedRequirements });
  const newbieRequirements = await status.public.callAny('steps_required', {
    level: 'newbie',
    user_id
  });
  snap({ newbieRequirements });

  const [userAchievedNewbie] = await status.public.callAny('user_achieved', {
    level: 'newbie',
    user_id
  });
  snap({ newbie: userAchievedNewbie });
  const [userAchievedAdvanced] = await status.public.callAny('user_achieved', {
    level: 'advanced',
    user_id
  });
  snap({ advanced: userAchievedAdvanced });
});

it('advanced part III', async () => {
  //

  const advanced = [
    ['apply_for_verifier'],
    ['approved_for_verifier'],
    ['create_action', 2]
  ];

  for (const [name, required_count = 1] of advanced) {
    await status.public.insertOne('level_requirements', {
      name,
      level: 'advanced',
      required_count
    });
  }

  const steps = [
    'agree_to_terms',
    'accept_cookies',
    'accept_privacy',
    'upload_profile_image',
    ...repeat('invite_users', 20),
    ...repeat('complete_action', 20),
    ...repeat('create_action', 10),
    ...repeat('apply_for_verifier', 1),
    ...repeat('approved_for_verifier', 1)
  ];
  for (const name of steps) {
    await status.public.insertOne('user_steps', {
      user_id,
      name
    });
  }

  const advancedRequirements = await status.public.callAny('steps_required', {
    level: 'advanced',
    user_id
  });
  snap({ advancedRequirements });
  const newbieRequirements = await status.public.callAny('steps_required', {
    level: 'newbie',
    user_id
  });
  snap({ newbieRequirements });

  const [userAchievedNewbie] = await status.public.callAny('user_achieved', {
    level: 'newbie',
    user_id
  });
  snap({ newbie: userAchievedNewbie });
  const [userAchievedAdvanced] = await status.public.callAny('user_achieved', {
    level: 'advanced',
    user_id
  });
  snap({ advanced: userAchievedAdvanced });
});
