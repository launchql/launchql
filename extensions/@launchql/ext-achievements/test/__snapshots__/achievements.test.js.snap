// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`advanced 1`] = `
Object {
  "advancedRequirements": Array [
    Object {
      "id": "[ID]",
      "level": "advanced",
      "name": "invite_users",
      "priority": 100,
      "required_count": 12,
    },
  ],
}
`;

exports[`advanced 2`] = `
Object {
  "newbieRequirements": Array [],
}
`;

exports[`advanced 3`] = `
Object {
  "newbie": Object {
    "user_achieved": true,
  },
}
`;

exports[`advanced 4`] = `
Object {
  "advanced": Object {
    "user_achieved": false,
  },
}
`;

exports[`advanced part II 1`] = `
Object {
  "advancedRequirements": Array [
    Object {
      "id": "[ID]",
      "level": "advanced",
      "name": "invite_users",
      "priority": 100,
      "required_count": 12,
    },
    Object {
      "id": "[ID]",
      "level": "advanced",
      "name": "complete_action",
      "priority": 100,
      "required_count": 5,
    },
    Object {
      "id": "[ID]",
      "level": "advanced",
      "name": "apply_for_verifier",
      "priority": 100,
      "required_count": 1,
    },
    Object {
      "id": "[ID]",
      "level": "advanced",
      "name": "create_action",
      "priority": 100,
      "required_count": 2,
    },
  ],
}
`;

exports[`advanced part II 2`] = `
Object {
  "newbieRequirements": Array [],
}
`;

exports[`advanced part II 3`] = `
Object {
  "newbie": Object {
    "user_achieved": true,
  },
}
`;

exports[`advanced part II 4`] = `
Object {
  "advanced": Object {
    "user_achieved": false,
  },
}
`;

exports[`advanced part III 1`] = `
Object {
  "advancedRequirements": Array [],
}
`;

exports[`advanced part III 2`] = `
Object {
  "newbieRequirements": Array [],
}
`;

exports[`advanced part III 3`] = `
Object {
  "newbie": Object {
    "user_achieved": true,
  },
}
`;

exports[`advanced part III 4`] = `
Object {
  "advanced": Object {
    "user_achieved": true,
  },
}
`;

exports[`newbie 1`] = `
Object {
  "advancedRequirements": Array [
    Object {
      "id": "[ID]",
      "level": "advanced",
      "name": "invite_users",
      "priority": 100,
      "required_count": 15,
    },
    Object {
      "id": "[ID]",
      "level": "advanced",
      "name": "complete_action",
      "priority": 100,
      "required_count": 12,
    },
  ],
}
`;

exports[`newbie 2`] = `
Object {
  "newbieRequirements": Array [
    Object {
      "id": "[ID]",
      "level": "newbie",
      "name": "upload_profile_image",
      "priority": 100,
      "required_count": 1,
    },
    Object {
      "id": "[ID]",
      "level": "newbie",
      "name": "complete_action",
      "priority": 100,
      "required_count": 2,
    },
  ],
}
`;

exports[`newbie 3`] = `
Object {
  "newbie": Object {
    "user_achieved": false,
  },
}
`;

exports[`newbie 4`] = `
Object {
  "advanced": Object {
    "user_achieved": false,
  },
}
`;
